### ISSUE TITLE
[Training] Initial Full Training Run & Monitoring Setup

### ISSUE THREAD

**Issue by fenghaolw** [fenghaolw/azul-helper#12]
- Integrate self-play and training into a continuous loop.
- Set initial hyperparameters (learning rate, batch size, MCTS simulations, replay buffer size).
- Implement basic logging/monitoring for loss values, game length, (if possible) simple score metrics.
- Run the first extended training session.
- Acceptance Criteria: The full self-play and training loop runs, and initial learning metrics are being logged.

**Comment by fenghaolw** [comment]
The fast training I can get right now is 25m one step, we collect states for 348 games (20000 states) and learner writes a checkpoint to a step. That means in 17 days we can finish 1000 steps.

Current key hyperparameters:

* `inference_threads = 1` - it seems like more threads actually harm the performance, at least on my mbp.
* `inference_batch_size = actors = 4` - actors self-play. Increasing this further also seems to have diminish returns on mbp. M4 pro has 10 cores so we have more CPUs to use, but the contention might kill the performance and does not increase the throughput further.
* `evaluators = 1`: I might see if increasing this helps.
* `replay_buffer_size = 100000`: This determines how many of the most recent game states are kept for training. 100,000 states is a bit small. It means agent will quickly discard data from older versions of itself. A larger buffer provides more diverse training data and can lead to more stable learning. So I might increase this in a later fresh training run, but I cannot do that now since the code will throw if it tries to load an existing buffer with different size.
* `inference_cache = 2000000`: This is crucial. Original slowness seems to purely come from a small cache size (1000). Increasing this to 1000000 causes the cache hit to raise from 43% to 69%, but we should aim for 80+% cache hit rate. We have enough RAM.
* `learning_rate = 0.001`: This might be suboptimal. Learning rate annealing (lower the rate in later steps) might be necessary.
* `train_batch_size = 256`: Larger batch sizes lead to more stable gradients for learners.

Progress so far:
```
[2025-06-06 07:24:36.915] Step: 1
[2025-06-06 07:24:36.915] Collected 10030 states from 238 games, 0.3 states/s; 0.1 states/(s*actor), game length: 42.1
[2025-06-06 07:24:36.915] Queue size: 0. Buffer size: 10030. States seen: 10030
[2025-06-06 07:25:12.150] Checkpoint saved: models/libtorch_alphazero_azul/checkpoint--1
[2025-06-06 07:25:12.150] Losses: policy: 1.6884, value: 0.7857, l2: 0.0777, sum: 2.5518
[2025-06-06 07:25:12.150] Cache size: 1000/1000: 100.0%, hits: 2642935, misses: 3495067, hit rate: 43.059%

[2025-06-06 09:38:40.754] Step: 2
[2025-06-06 09:38:40.754] Collected 10015 states from 348 games, 6.4 states/s; 1.6 states/(s*actor), game length: 28.8
[2025-06-06 09:38:40.754] Queue size: 0. Buffer size: 20045. States seen: 20045
[2025-06-06 09:39:43.761] Checkpoint saved: models/libtorch_alphazero_azul/checkpoint--1
[2025-06-06 09:39:43.761] Losses: policy: 1.6864, value: 0.7546, l2: 0.0595, sum: 2.5005
[2025-06-06 09:39:43.761] Cache size: 1000000/1000000: 100.0%, hits: 4140361, misses: 1830377, hit rate: 69.344%
```

Action items:
- [ ] Continue monitoring cache hit and losses
- [ ] Investigate learning rate annealing and if I need to adjust learning rate manually
- [ ] Create an agent using checkpoint models and see its strength

**Comment by fenghaolw** [comment]
I am excited. Now with just 10 steps of training (a few hours), the agent is already beating minimax and random rollout consistently.

```
FINAL RANKINGS:
  1. AlphaZero_MCTS_400: 81.7% win rate, 10.7 avg score
  2. MCTS_1000: 43.3% win rate, 8.2 avg score
  3. Minimax_D4: 25.0% win rate, 10.9 avg score

HEAD-TO-HEAD RESULTS:
Minimax_D4 vs MCTS_1000: 40.0% win rate (12/30 games)
Minimax_D4 vs AlphaZero_MCTS_400: 10.0% win rate (3/30 games)
MCTS_1000 vs AlphaZero_MCTS_400: 26.7% win rate (8/30 games)
```

**Comment by fenghaolw** [comment]
Initially, I was overly optimistic. A [significant bug](https://github.com/fenghaolw/azul-helper/commit/b3fa5e5dcbf99e1968f378b4c0872b8a7ec3d5d0) in the game implementation, specifically the factory count being incorrectly set to one, drastically reduced the action space. The AlphaZero agent learned to "master" this flawed game, leading to seemingly impressive performance in the broken scenario.

However, after fixing the bug and re-training with different sets of hyperparameters, the AlphaZero agent consistently lost to minimax-d3, an AI not known for its strength and lacking a well-designed value function. The losses were severe, with the AlphaZero agent losing 100% of games by a large margin (0-60 score).

Here is what Gemini told me:

> This is a classic and difficult problem in self-play learning. You have successfully created an agent that is trapped in a stable but flawed equilibrium.
> 
> ## The Diagnosis: The "Island Civilization" Problem
> Imagine your agent is a civilization that grew up on an isolated island.
> * Initial Discovery: From nothing, it discovered a way to live. This was your "rushdown" strategy of short, 20-move games. On their island, this is the most efficient way to survive.
> * Perfection: For 12 hours and hundreds of thousands of games, your civilization has done nothing but perfect this one way of life. They have become incredibly, optimally efficient at their island's meta-game. Their "loss" decreases because they get better and better at their own system.
> * The Outsider: Minimax_D3 arrives from the mainland. It doesn't play by the island's rules. It uses tools and strategies (like building for end-game bonuses, managing a 50-move game) that the islanders have never conceived of.
> * The Collapse: The islanders' hyper-specialized strategy is useless against this new context. They are completely wiped out.
> This is what has happened to your agent. Increasing the temperature_drop was like encouraging the islanders to explore more of their own island. It created local diversity, but it never introduced the outside knowledge required to break the flawed equilibrium.
> 
> ## The Action Plan: The Unavoidable Solutions
> You have exhausted the "pure self-play" tuning options. The problem is not with the hyperparameters anymore; it's with the fundamental training methodology. You must now force the agent to learn from outside knowledge and change its core motivation.
> I strongly recommend starting a new run from scratch after deleting the old data. The current model and replay buffer are "contaminated" with the failed strategy and will prevent the agent from learning effectively.
> Here are the two critical changes you need to make for the new run. I would recommend implementing both.
> ### 1. Introduce "Sparring Partners" (Break the Isolation)
> The agent must be forced to play against different strategies. You need to modify your actor threads so they don't exclusively play against themselves.
> Action: Modify your actor's game-playing loop. Implement a schedule where, for example:
> * 70% of games are normal self-play (Agent vs. latest Agent).
> * 30% of games are against a pool of sparring partners, chief among them being Minimax_D3.
> **Why it Works:** The replay buffer will be seeded with devastating losses. The agent's value network will be forced to learn that the board states resulting from its "rushdown" strategy are, in fact, catastrophic -1 outcomes when faced with a standard opponent. This feedback is brutal, direct, and the fastest way to break the echo chamber.
> ### 2. Modify the Reward Signal (Change the Goal)
> The agent currently sees no difference between winning by 1 point and winning by 40. We need to teach it to be "greedy."
> Action: Change the terminal reward of the game from a binary +1/-1 to a normalized score differential.
> * Calculate score_diff = your_score - opponent_score.
> * Normalize it into the [-1, 1] range. The tanh function is perfect for this: reward = tanh(score_diff / scaling_factor). A scaling factor of 40 or 50 is a good starting point.
> **Why it Works:** This fundamentally changes the agent's objective. It is no longer trying to just win; it is trying to crush the opponent. It will be incentivized to learn the complex, high-scoring strategies required for a 60-0 victory because that provides a much higher reward signal than a 20-15 win. This directly combats the low-scoring meta it has developed.

I have implemented (2) in https://github.com/fenghaolw/azul-helper/commit/7328087d7e290c4afe4452e287b4978bc0b5586c and restarted the training. We will see how that goes. Meanwhile I will look into (1) which might be tricky to implement in the existing open_spiel implementation.

**Comment by fenghaolw** [comment]
With the reward changes and ~50 steps, now I have a decent AI

```
=== EVALUATION RESULTS ===
Agent 1: AlphaZero_MCTS_3200
Agent 2: Minimax_D3
Games played: 10
AlphaZero_MCTS_3200 wins: 9
Win rate: 90.000%
Statistical significance: Yes
95% Confidence interval: [71.406%, 100.000%]
```

It defeats minimax consistently. When I played with it, it plays reasonably well. I probably still can win 30% of the games, but since we are only 50 steps in (AlphaZero usually requires 1000+ steps to have a strong performance), I am already happy.

Losses have been steadily decreasing

![Image](https://github.com/user-attachments/assets/99d1bd81-451b-4147-9d48-c8cd347905cf)
